src/lib/client.js

```diff
@@ -26,18 +30,93 @@ module.exports = class BtpClient {
     this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
   }
 
+  tryToOpenWebSocketClient () {
+    return new Promise((resolve, reject) => {
+      const ws = new WebSocket(this._wsUri, {
+        perMessageDeflate: false
+      })
+      console.log('try to open', this._wsUri)
+      let hasBeenOpen = false
+      ws.on('open', () => {
+        console.log('open!')
+        hasBeenOpen = true
+        resolve(ws)
+      })
+      ws.on('error', (err) => {
+        if (hasBeenOpen) {
+          console.error('WebSocket error', err)
+        } else {
+          reject(err)
+        }
+      })
+      ws.on('close', () => {
+        console.log('close!')
+        // automatically reconnect if server reboots
+        if (hasBeenOpen && !ws.shouldClose) {
```

Comment:
can hasBeenOpen ever be false when 'close' is emitted? If so, it would cause this promise to hang forever.

Reply:
i'll check

Reply:
there are three event orders:
open -> close
open -> error -> close
error -> close (hasBeenOpen is false)

the promise wouldn't hang because close is either preceded by open (the promise was resolved) or by error (the promise was rejected)